registry/registry.go

```diff
@@ -0,0 +1,93 @@
+package registry
+
+import (
+	"context"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/tcsc/levity/task"
+)
+
+// An empty task ID to return on error
+var nullTaskHandle = ""
+
+// handleGeneratorFn represents an arbitrary function that can be called to
+// generate a candidate TaskHandle value.
+type handleGeneratorFn func() string
+
+// Registry models a simple task registration system. For a simple exercise
+// like this we are just keeping the task information in memory. This would
+// obviously require somthing more durable in a production server.
+type Registry struct {
+	lock           sync.RWMutex
+	db             map[string]*task.Task
+	generateHandle handleGeneratorFn
+}
+
+func handleFromUUID() string {
+	return uuid.New().String()
+}
+
+// New creates and returns an initialised, ready-to-use task registry
+func New() *Registry {
+	return newWithIDGenerator(handleFromUUID)
+}
+
+func newWithIDGenerator(gen handleGeneratorFn) *Registry {
+	return &Registry{
+		db:             make(map[string]*task.Task),
+		generateHandle: gen,
+	}
+}
+
+// Lookup attempts to find a task from its ID. Returns nil if no such task exists.
+func (registry *Registry) Lookup(handle string) *task.Task {
+	registry.lock.RLock()
+	defer registry.lock.RUnlock()
+
+	if t, exists := registry.db[handle]; exists {
+		return t
+	}
+
+	return nil
+}
+
+// Register attempts to register a task with the registry. Keeps trying until
+// it succeeds in generateing a unique ID, or the supplied context expires.
+func (registry *Registry) Register(ctx context.Context, t *task.Task) (string, error) {
+	// For the purposes of this exercise we're using a stringified UUID as
+	// the task ID. This should be statistically unique - certainly we
+	// should not expect to see a collision in the lifetime of the server.
+	//
+	// To handle the remote possibility of a collision, we keep trying until
+	// we have a unique ID registered in the database. To handle the even more
+	// remote possibility of repeated collisions, we will only keep trying
+	// until the supplied context expires
+	for ctx.Err() == nil {
```

Comment:
My basic idea was that the function would keep trying to generate a unique ID until the context completed, rather than just using an arbitrary number of retries. I wasn't sure how to detect that a context was closed, and polling the error  seemed like a good idea at the time. 

I have replaced with a `select` over `ctx.Done()` that has a `default` case, so we don't actually block on the channel.

Reply:
Refactored away, in any case